kmk/kmk_keyboard.py

```diff
@@ -90,223 +87,338 @@ def __repr__(self):
+            self._hid_helper = USBHID
+        elif self.hid_type == HIDModes.BLE:
+            self._hid_helper = BLEHID
         else:
-            self.is_target = True
+            self._hid_helper = AbstractHID
+        self._hid_helper = self._hid_helper()
 
-        if self.uart_pin is not None:
-            self.uart = self.init_uart(self.uart_pin)
-
-        if self.rgb_pixel_pin:
-            self.pixels = rgb.RGB(self.rgb_config, self.rgb_pixel_pin)
-            self.rgb_config = None  # No longer needed
-            self.pixels.loopcounter = 0
-        else:
-            self.pixels = None
-
-        if self.led_pin:
-            self.led = led.led(self.led_pin, self.led_config)
-            self.led_config = None  # No longer needed
-        else:
-            self.led = None
-
-        self.matrix = self.matrix_scanner(
+    def _init_matrix(self):
+        self.matrix = MatrixScanner(
             cols=self.col_pins,
             rows=self.row_pins,
             diode_orientation=self.diode_orientation,
             rollover_cols_every_rows=getattr(self, 'rollover_cols_every_rows', None),
         )
 
-        # Compile string leader sequences
-        for k, v in self.leader_dictionary.items():
-            if not isinstance(k, tuple):
-                new_key = tuple(KC[c] for c in k)
-                self.leader_dictionary[new_key] = v
+        return self
 
-        for k, v in self.leader_dictionary.items():
-            if not isinstance(k, tuple):
-                del self.leader_dictionary[k]
+    # Only one GC to allow for extentions to have room. There are random memory allocations
+    # issues due to some devices not properly cleaning memory on reset
+    gc.collect()
 
-        while True:
-            if self.split_type is not None and self.is_target:
-                update = self._receive_from_initiator()
-                if update is not None:
-                    self._handle_matrix_report(update)
+    def go(self, hid_type=HIDModes.USB, **kwargs):
+        self.hid_type = hid_type
 
-            update = self.matrix.scan_for_changes()
+        self._init_sanity_check()
+        self._init_coord_mapping()
+        self._init_hid()
 
-            if update is not None:
-                if self.is_target:
-                    self._handle_matrix_report(update)
-                else:
-                    # This keyboard is a initiator, and needs to send data to target
-                    self._send_to_target(update)
+        for ext in self.extensions:
+            try:
+                ext.during_bootup(self)
+            except Exception:
+                print('Failed to load extention', ext)
+
+        self._init_matrix()
 
-            if self._state.hid_pending:
+        self._print_debug_cycle(init=True)
+
+        while True:
+            self.state_changed = False
+
+            for ext in self.extensions:
+                try:
+                    self._handle_matrix_report(ext.before_matrix_scan(self))
```

Comment:
Can we store and handle this and not just blind fire assuming it's a matrix report? I believe this would save power as well as allow split keyboards to report matrix updates to extensions from all parts, and not just what it's on. A simple example of why one would need to do this is OLED's timeout, calculating WPM, and several other ideas come to mind.